File: bench/src/go/map_kernel.go
// Equal-workload kernel — see bench/src/c/map_kernel.c. Same open-addressing map, same hash, same
// fixed prealloc, no stdlib map (Go's builtin `map` is measured in the `map` row instead).
package main

import "os"

type IntMap struct {
	keys []int32
	vals []int64
	used []uint8
	cap  int32
}

func newIntMap(cap int32) *IntMap {
	return &IntMap{keys: make([]int32, cap), vals: make([]int64, cap), used: make([]uint8, cap), cap: cap}
}

func (m *IntMap) slotFor(k int32) int32 {
	x := uint64(uint32(k)) * 2654435761 // hash
	return int32(x % uint64(m.cap))
}

func (m *IntMap) put(k int32, v int64) {
	i := m.slotFor(k)
	for m.used[i] != 0 {
		if m.keys[i] == k {
			m.vals[i] = v
			return
		}
		i++
		if i >= m.cap {
			i = 0
		}
	}
	m.keys[i] = k
	m.vals[i] = v
	m.used[i] = 1
}

func (m *IntMap) get(k int32) int64 {
	i := m.slotFor(k)
	for m.used[i] != 0 {
		if m.keys[i] == k {
			return m.vals[i]
		}
		i++
		if i >= m.cap {
			i = 0
		}
	}
	return 0
}

func main() {
	const N int64 = 100000
	const PASSES int64 = 10
	m := newIntMap(262144) // > N / 0.75, power of 2
	for i := int64(0); i < N; i++ {
		m.put(int32(i), i*2)
	}
	var sum uint64 = 0
	for p := int64(0); p < PASSES; p++ {
		for i := int64(0); i < N; i++ {
			k := int32((i * 2654435761) % N)
			sum += uint64(m.get(k))
		}
	}
	os.Exit(int(sum % 256))
}

File: bench/src/js/map_kernel.js
// Equal-workload kernel — see bench/src/c/map_kernel.c. Same open-addressing map, same hash, same
// fixed prealloc, no stdlib Map (JS's `Map` is measured in the `map` row instead). Typed arrays keep
// the element types honest; every intermediate stays < 2^53, so float64 arithmetic is exact.
const CAP = 262144;               // > N / 0.75, power of 2
const N = 100000, PASSES = 10;

const keys = new Int32Array(CAP);
const vals = new Float64Array(CAP);   // int64 values here peak at 199 998 — exact in float64
const used = new Uint8Array(CAP);

function slotFor(k) {
    return ((k >>> 0) * 2654435761) % CAP;   // hash; < 2^53, exact
}

function put(k, v) {
    let i = slotFor(k);
    while (used[i] !== 0) {
        if (keys[i] === k) { vals[i] = v; return; }
        i++; if (i >= CAP) i = 0;
    }
    keys[i] = k; vals[i] = v; used[i] = 1;
}

function get(k) {
    let i = slotFor(k);
    while (used[i] !== 0) {
        if (keys[i] === k) return vals[i];
        i++; if (i >= CAP) i = 0;
    }
    return 0;
}

for (let i = 0; i < N; i++) put(i, i * 2);
let sum = 0;
for (let p = 0; p < PASSES; p++)
    for (let i = 0; i < N; i++) sum += get((i * 2654435761) % N);
process.exit(sum % 256);
